axlearn/common/trainer_config_modifier.py

```diff
@@ -0,0 +1,150 @@
+            metric_accumulator=self._metric_accumulator,
+        )
+        return cfg
+
+
+class RematSpecModifier(ConfigModifier):
+    """Update the remat policies for specified modules."""
+
+    @config_class
+    class Config(ConfigModifier.Config):
+        remat_policies: Optional[Dict[str, RematSpec]] = None
+
+    def __init__(self, cfg: Config):
+        super().__init__(cfg)
+        cfg = self.config
+        self._remat_policies = cfg.remat_policies
+
+    def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        """Update the remat policy for the specified modules.
+
+        Args:
+            cfg (SpmdTrainer.Config): the trainer config to be modified.
+
+        Raises:
+            ValueError: the target module is not found.
+            ValueError: the remat_spec attribute is not found.
+
+        Returns:
+            The modified trainer config.
+        """
+        if self._remat_policies is None:
+            return cfg
+
+        for module_name, remat_spec in self._remat_policies.items():
+            # Here we assume x.y.z format.
+            # One example would be model.decoder.transformer.layer.
+            target_modules = module_name.split(".")
+            curr_module = cfg
+            for target_module in target_modules:
+                if not hasattr(curr_module, target_module):
+                    raise ValueError(f"{target_module} is not found in {curr_module}.")
+                curr_module = getattr(curr_module, target_module)
+            # Here we assume all modules have remat_spec attribute.
+            if not hasattr(curr_module, "remat_spec"):
+                raise ValueError(f"{curr_module} does not have remat_spec attribute")
+            curr_module.remat_spec = remat_spec
+        return cfg
+
+
+class MeshShapeModifier(ConfigModifier):
+    """Update the mesh_shape for the trainer config."""
+
+    @config_class
+    class Config(ConfigModifier.Config):
+        mesh_shape: Optional[Union[MeshShape, HybridMeshShape]] = None
+
+    def __init__(self, cfg: Config):
+        super().__init__(cfg)
+        cfg = self.config
+        self._mesh_shape = cfg.mesh_shape
+
+    def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        """Overwrite the mesh shape.
+
+        Args:
+            cfg (SpmdTrainer.Config): the trainer config to be modified.
+
+        Returns:
+            The modified trainer config.
+        """
+        cfg.mesh_shape = self._mesh_shape
+        return cfg
+
+
+class ChainConfigModifier(ConfigModifier):
+    """Chain multiple config modifiers together."""
+
+    @config_class
+    class Config(ConfigModifier.Config):
+        config_modifiers: Required[List[ConfigOr[ConfigModifier]]] = REQUIRED
+
+    def __init__(self, cfg: Config):
+        super().__init__(cfg)
+        cfg = self.config
+        self._config_modifiers = [
+            cfg_modifier.instantiate() for cfg_modifier in cfg.config_modifiers
```

Comment:
Since we specify the config modifier as `ConfigOr[ConfigModifier]`, we should use
```suggestion
            maybe_instantiate(cfg_modifier) for cfg_modifier in cfg.config_modifiers
```